maths/binary_exponentiation.py

```diff
@@ -1,48 +1,148 @@
+    3
+    >>> binary_exp_iterative(3, 0)
+    1
+    >>> binary_exp_iterative(3, -1)
+    Traceback (most recent call last):
+        ...
+    ValueError: Exponent must be a non-negative integer
+    """
+    if exponent < 0:
+        raise ValueError("Exponent must be a non-negative integer")
+
+    res = 1
+    while exponent > 0:
+        if exponent & 1:
+            res *= base
+
+        base *= base
+        exponent >>= 1
+
+    return res
+
+
+def binary_exp_mod_recursive(base: int, exponent: int, modulus: int) -> int:
+    """
+    >>> binary_exp_mod_recursive(3, 4, 5)
+    1
+    >>> binary_exp_mod_recursive(7, 13, 10)
+    7
+    >>> binary_exp_mod_recursive(7, -1, 10)
+    Traceback (most recent call last):
+        ...
+    ValueError: Exponent must be a non-negative integer
+    >>> binary_exp_mod_recursive(7, 13, 0)
+    Traceback (most recent call last):
+        ...
+    ValueError: Modulus must be a positive integer
+    """
+    if exponent < 0:
+        raise ValueError("Exponent must be a non-negative integer")
+    if modulus <= 0:
+        raise ValueError("Modulus must be a positive integer")
+
+    if exponent == 0:
         return 1
 
-    elif n % 2 == 1:
-        return binary_exponentiation(a, n - 1) * a
+    if exponent % 2 == 1:
+        return (binary_exp_mod_recursive(base, exponent - 1, modulus) * base) % modulus
+
+    r = binary_exp_mod_recursive(base, exponent // 2, modulus)
+    return (r * r) % modulus
+
+
+def binary_exp_mod_iterative(base: int, exponent: int, modulus: int) -> int:
+    """
+    >>> binary_exp_mod_iterative(3, 4, 5)
+    1
+    >>> binary_exp_mod_iterative(7, 13, 10)
+    7
+    >>> binary_exp_mod_iterative(7, -1, 10)
+    Traceback (most recent call last):
+        ...
+    ValueError: Exponent must be a non-negative integer
+    >>> binary_exp_mod_iterative(7, 13, 0)
+    Traceback (most recent call last):
+        ...
+    ValueError: Modulus must be a positive integer
+    """
+    if exponent < 0:
+        raise ValueError("Exponent must be a non-negative integer")
+    if modulus <= 0:
+        raise ValueError("Modulus must be a positive integer")
+
+    res = 1
+    while exponent > 0:
+        if exponent & 1:
+            res = ((res % modulus) * (base % modulus)) % modulus
+
+        base *= base
+        exponent >>= 1
 
-    else:
-        b = binary_exponentiation(a, n // 2)
-        return b * b
+    return res
 
 
 if __name__ == "__main__":
     import doctest
 
     doctest.testmod()
 
-    try:
-        BASE = int(float(input("Enter Base : ").strip()))
-        POWER = int(input("Enter Power : ").strip())
-    except ValueError:
-        print("Invalid literal for integer")
+    BASE = int(input("Enter base: ").strip())
+    POWER = int(input("Enter power: ").strip())
 
-    RESULT = binary_exponentiation(BASE, POWER)
-    print(f"{BASE}^({POWER}) : {RESULT}")
+    RESULT = binary_exp_recursive(BASE, POWER)
+    print(f"{BASE}^({POWER}): {RESULT}")
```

Comment:
OPTIONAL: Adding a `timeit` benchmark would be a nice addition.  Visitors love to see the side-by-side bakeoff.